### PR TITLE
RSDK-1656: Verify UR5 driver connectivity logic

### PR DESCRIPTION
**Pull request by nfranczak**
Apart from the software test, this PR was also tested on a real UR5E.

The test on hardware ensured that changing the speed value actually resulted in speed changes and that this was done without the arm instance being re-created.

The speed change was visually confirmed.
It was also confirmed that the speed change was implemented without the arm being re-created. This was verified through specific server error logs not showing up.

Before this PR, what would happen is that a new arm instance is started FIRST, then arm instances are swapped in the `.actual` member of the reconfigurableArm object, and only then is the old instance Close()ed.
This would result in the following logs:
```
2023-02-07T14:24:28.828-0500	WARN	robot_server	impl/local_robot.go:688	using legacy constructor	{"subtype": "rdk:component:arm", "model": "rdk:builtin:ur5e"}
2023-02-07T14:24:28.873-0500	ERROR	robot_server	universalrobots/ur.go:233	dashboard reader failed	{"error": "context canceled"}
2023-02-07T14:24:28.873-0500	ERROR	robot_server	universalrobots/ur.go:267	reader failed	{"error": "context canceled"}
2023-02-07T14:24:28.873-0500	ERROR	robot_server	arm/arm.go:330	error closing old	{"error": "context canceled"}
2023-02-07T14:24:28.873-0500	ERROR	robot_server	arm/arm.go:295	error closing old	{"error": "context canceled"}
```

now, when changing the speed values on the UI, the logs above do not show up.

### CODE REVIEW

File: components/arm/universalrobots/ur.go
Comment (biotinker) [inline components/arm/universalrobots/ur.go:119]: This will almost certainly fail terribly; if `host` changes it's probably worth doing the full reconfigure.

**Comment by biotinker** [comment]
Can we please test this on a real arm and observe speed changes happening without the arm being recreated?

File: components/arm/universalrobots/ur.go
Comment (raybjork) [inline components/arm/universalrobots/ur.go:122]: pretty sure this can just become 
```
ua.urHostedKinematics != newCfg.ArmHostedKinematics
ua.speed != newCfg.Speed
```

File: components/arm/universalrobots/ur.go
Comment (raybjork) [inline components/arm/universalrobots/ur.go:115]: [nit] pretty sure doing is this way is stylistically preferred
`if newCfg, ok := c.ConvertedAttributes.(*AttrConfig); ok {`

**Review by raybjork — APPROVED** [review]

**Review by raybjork — APPROVED** [review]

File: components/arm/universalrobots/ur.go
Comment (biotinker) [inline components/arm/universalrobots/ur.go:114]: Note that this will always alter the existing struct even if it turns out that we do not want to keep it. This could cause undesirable behavior.

Consider instead structuring it such that the things that trigger a full reconfigure, do so explicitly, and the struct is not altered unless we definitely want to alter it.

```
		if ua.host != newCfg.Host {
			return config.Reconfigure
		}
		ua.speed = newCfg.Speed
		ua.urHostedKinematics = newCfg.ArmHostedKinematics
		return config.None
```
or

```
		if ua.host == newCfg.Host {
			ua.speed = newCfg.Speed
			ua.urHostedKinematics = newCfg.ArmHostedKinematics
			return config.None
		}
```
I consider the first one slightly more readable and extensible.

**Comment by biotinker** [comment]
> now, when changing the speed values on the UI, the logs above do not show up.

Can you confirm that the speed of the arm did in fact change?

**Review by biotinker — APPROVED** [review]
This LGTM assuming completion of real-world testing showing that this is doing what is expected

**Comment by nfranczak** [comment]
> > now, when changing the speed values on the UI, the logs above do not show up.
> 
> Can you confirm that the speed of the arm did in fact change?

@biotinker , yep. I confirmed this visually.

File: components/arm/universalrobots/ur5e_test.go
Comment (npmenard) [inline components/arm/universalrobots/ur5e_test.go:474]: nit: more telling name

File: components/arm/universalrobots/ur5e_test.go
Comment (npmenard) [inline components/arm/universalrobots/ur5e_test.go:465]: nit: more telling name

File: components/arm/universalrobots/ur5e_test.go
Comment (npmenard) [inline components/arm/universalrobots/ur5e_test.go:494]: you don't need the if  statement if the above test pass

File: components/arm/universalrobots/ur5e_test.go
Comment (npmenard) [inline components/arm/universalrobots/ur5e_test.go:503]: nit: can just do `test.That(t,obj.UpdateAction(&lastCfg), test.ShouldEqual, config.Reconfigure)`

File: components/arm/universalrobots/ur5e_test.go
Comment (npmenard) [inline components/arm/universalrobots/ur5e_test.go:508]: use defer here, if a test fails we don't want to leave lingering resources

File: components/arm/universalrobots/ur5e_test.go
Comment (npmenard) [inline components/arm/universalrobots/ur5e_test.go:485]: might not be possible but if there is a way to create the arm by just using the URArm struct that would make your code easier since you don't the arm communication emulation layer

File: components/arm/universalrobots/ur5e_test.go
Comment (npmenard) [inline components/arm/universalrobots/ur5e_test.go:489]: you should also wrap your arm with reconfigurable arm so you test the codepath that will be executed during reconfigure.

**Review by npmenard — CHANGES_REQUESTED** [review]
LGTM, couple a comments that should be addressed before review

File: components/arm/universalrobots/ur5e_test.go
Comment (nfranczak) [reply components/arm/universalrobots/ur5e_test.go:485]: I do not think it is possible to do that, but I defer to @biotinker for confirmation.

File: components/arm/universalrobots/ur5e_test.go
Comment (nfranczak) [inline components/arm/universalrobots/ur5e_test.go:504]: @npmenard added wrap + testing here

File: components/arm/universalrobots/ur5e_test.go
Comment (biotinker) [inline components/arm/universalrobots/ur5e_test.go:487]: Won't this fail if this test does not pass in the connection timeout period?

I think this will also work if you do:
```
attrs, ok := cfg.ConvertedAttributes.(*AttrConfig)
test.That(ok is true)
&URArm{
	speed:                    attrs.Speed,
	urHostedKinematics:       attrs.ArmHostedKinematics,
	host:                     attrs.Host,
}
```
and this will avoid attempts to make network calls

Comment (nfranczak) [reply components/arm/universalrobots/ur5e_test.go:487]: this is a cool solution to get around making network calls

**Review by biotinker — APPROVED** [review]
LGTM!

**Review by spieswl — APPROVED** [review]
Looks good! Nice description of the manual testing.

**Pull request merged**
